File: ml_data/data_methods.py
# MIT license

# Copyright (c) 2024 ???

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import logging
import mailbox
import pandas as pd
import requests

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s"
)

def get_messages(list_name):
    """
    Download and save mbox file from Apache Arrow mailing list archive.
    """
    url = f"https://lists.apache.org/api/mbox.lua?list={list_name}&domain=arrow.apache.org"
    logging.info(f"Starting download of mbox file from Apache Arrow {list_name} mailing list.")

    try:
        resp = requests.get(url)
        resp.raise_for_status()
        with open(f"{list_name}_ml.mbox", mode="wb") as file:
            file.write(resp.content)
        logging.info(f"Successfully downloaded and saved {list_name} mbox file.")
    except requests.exceptions.RequestException as e:
        logging.error(f"Failed to download {list_name} mbox file: {e}")
        raise

def get_all(component):
    """
    Get user mailing list threads and email subject from the last
    month that are labelled with a particular component.

    This method needs mbox file to be saved to ./'user_ml.mbox'.

    Parameters
    ----------
    component : string
        Python or R.

    Returns
    -------
    issues : pd.DataFrame
        Pandas data frame with mailing list information.
    """
    logging.info(f"Processing mbox file for component: {component}")

    if component == "R":
        component = "[R]"

    try:
        email_list = [
            (message["Date"], message["Subject"], message["Thread-Topic"])
            for message in mailbox.mbox("user_ml.mbox")
            if component.lower() in (message["Subject"] or "").lower()
        ]

        logging.info(f"Found {len(email_list)} messages related to component: {component}")

        df = pd.DataFrame(email_list, columns=["date", "subject", "thread"])

        # Add url_title column with href link to the search result of the email
        df["url_title"] = df.subject.str.replace(" ", "%20", regex=False)
        df["url_title"] = (
            '<a target="_blank" href="https://lists.apache.org/list?user@arrow.apache.org:lte=1M'
            + df["url_title"]
            + '">'
            + df["subject"]
            + "</a>"
        )

        logging.info("Successfully created DataFrame with mailing list information.")
        return df[["date", "url_title"]]

    except Exception as e:
        logging.error(f"Error while processing mbox file: {e}")
        raise

File: stack_overflow_data/data_methods.R
# MIT license

# Copyright (c) 2024 ???

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

library(DT)
library(dplyr)
library(glue)
library(gt)
library(httr)
library(lubridate)
library(tidyr)

get_data <- function(api_name, url){

  api_data <- httr::GET(url)
  out <- content(api_data, as = "text") %>%
    jsonlite::fromJSON()

  out
}

#' Get questions from Stack Overflow containing the specified tags
#'
#' @param tags Character vector of tags
#' @examples
#' get_questions(c("apache-arrow", "r"))
#' get_questions(c("apache-arrow", "python"))
#' get_questions("pyarrow")
#'
get_questions <- function(tags){

  tag_string <- paste(tags, collapse = "%3B")
  url <- glue("https://api.stackexchange.com/2.3/questions?order=desc&sort=activity&tagged={tag_string}&site=stackoverflow&filter=!-nt6H9OZ4WW*msaSa)YvngdWhKQ).R9VfXkayFbhnB61(g5UUJbH7f")

  out <- get_data(
    api_name = "questions",
    url = url
  )

  out
}

get_comments <- function(question_ids){

  comments_content <- get_data(
    api_name = "comments",
    url = paste0(
      "https://api.stackexchange.com/2.3/questions/",
      paste(question_ids, collapse = ";"),
      "/comments?order=desc&sort=creation&site=stackoverflow&filter=!-).qJXDT0Z5I"
    )
  )

  if (nrow(comments_content$items) == 0) {
    return(NULL)
  }

  comments_content$items %>%
    tidyr::unnest(owner) %>%
    select(display_name, account_id, score, post_id, comment_id, creation_date)

}

get_answers <- function(question_ids){

  answers_content <- get_data(
    api_name = "answers",
    url = paste0(
      "https://api.stackexchange.com/2.3/questions/",
      paste(question_ids, collapse = ";"),
      "/answers?order=desc&sort=activity&site=stackoverflow&filter=!9Rgp29w2U"
    )
  )

  if (nrow(answers_content$items) == 0) {
    return(NULL)
  }

  answers_content$items %>%
    tidyr::unnest(owner) %>%
    select(display_name, account_id, score, is_accepted, question_id, creation_date)
}

adjust_data <- function(questions_content){

  questions <- questions_content$items %>%
    mutate(last_activity_date = as_datetime(last_activity_date)) %>%
    mutate(retrieved = now())

  # retrieve comments
  comments <- get_comments(questions$question_id)

  # get counts of comments
  reply_counts <- comments %>%
    select(question_id = post_id) %>%
    group_by(question_id) %>%
    summarise(comments = n())

  answers <- get_answers(questions$question_id)
  answer_counts <- answers %>%
    select(question_id) %>%
    group_by(question_id) %>%
    summarise(answers = n())

  answer_accepted <- answers %>%
    filter(is_accepted) %>%
    select(question_id, is_accepted)

  # add in raw data, reply counts, and whether an answer has been accepted
  left_join(questions, reply_counts, by = "question_id") %>%
    left_join(answer_counts, by = "question_id") %>%
    left_join(answer_accepted, by = "question_id") %>%
    replace_na(list(is_accepted = FALSE, comments = 0, answers = 0)) %>%
      select(link, title, last_activity_date, comments, answers, accepted_answer = is_accepted) %>%
      mutate(days_since_last_activity = round(as.numeric(as.duration(interval(last_activity_date, now())), "days"))) %>%
      mutate(issue = paste0('<a href="',link,'" target="_blank">', title , '</a>'))
}

gt_show_questions <- function(data_adjusted){
  display_data <- data_adjusted %>%
    mutate(Question = paste0("[", title, "](", link, ")")) %>%
    select(Question, Answers = answers, `Accepted?` = accepted_answer,
           Comments = comments, `Days since activity` = days_since_last_activity)

  no_answer_rows <- which(!data_adjusted$accepted_answer)

  tbl <- display_data %>%
    gt() %>%
    fmt_markdown(columns = Question) %>%
    cols_width(
      Question ~ pct(50),
      Answers ~ pct(12),
      `Accepted?` ~ pct(12),
      Comments ~ pct(12),
      `Days since activity` ~ pct(14)
    ) %>%
    tab_options(
      table.width = pct(100),
      container.height = px(400),
      container.overflow.y = TRUE
    )

  if (length(no_answer_rows) > 0) {
    tbl <- tbl %>%
      tab_style(
        style = cell_fill(color = "#fbcd9989"),
        locations = cells_body(rows = no_answer_rows)
      )
  }

  tbl
}
